### DISCUSSION TITLE
mail notifications via postfix

### DISCUSSION THREAD

**Discussion by murdocklawless** [fail2ban/fail2ban#2954] [General]
hi,

my server is using postfix.

if I make necessary definitions in stmp.py like below;

def __init__(
		self, jail, name, host="localhost:25", user=user@domain.com, password=mailpassword,
		sendername="Fail2Ban", sender="fail2ban", dest="user@domain.com", matches=None):

and add lines to jail.local like below;

[DEFAULT]
mta = postfix
action = %(action_mwl)s

are these configs enough to send notification mails?

**Comment by sebres** [comment]
> if I make necessary definitions in stmp.py like below

You don't need to set it in `stmp.py` directly. This is an action (as many others) and you can supply every parameter to the action directly in jail:
```ini
[some-jail]
action = %(action_)s
         stmp.py[user="user@domain.com", dest="user@domain.com", ...]
```

> `action = %(action_mwl)s`
> are these configs enough to send notification mails?

Substitution `action_mwl` has nothing to do with stmp.py and uses another action `%(mta)s-whois-lines` - https://github.com/fail2ban/fail2ban/blob/884cbbd6e17dd97a87a5c6f7527c0512698ea5d2/config/jail.conf#L220-L221
and default `mta` is sendmail (you can change it setting `mta = mail` in jail or default section of your jail.local).

There is no such action postfix-whois-lines, so you have either to select other mta or write your own action (but sendmail or mail must work with postfix also). Take a look into the actions for the command line which will be used.

For other parameters that action may need take a look in `jail.conf` or correcponding action `action.d/*-whois-lines.conf`.
For instance for the `action_mwl` it is enough to set (overwrite) sender and destemail variables that will be interpolated in action substitution:
```ini
[DEFAULT]
sender = ...
destemail = ...
```

**Comment by murdocklawless** [comment]
I made a config as your advice below and service did not start after restart, it's saying;

ERROR Failed during configuration: While reading from '/etc/fail2ban/jail.local' [**_line 63_**]: option 'action' in section 'name
ERROR Async configuration of server failed

my jail.local config;
```
[DEFAULT]
mta = mail
sender = report@xxxxxx.com
sendername = Fail2ban

[named-refused-udp]
enabled   = true
port      = domain,953
protocol  = udp
filter    = named-refused
action    = iptables-multiport[name=named-udp, port="domain,953", protocol=udp]
logpath   = /var/log/syslog
action    = %(action_)s  **<--------------- _line 63_**
            stmp.py[host="localhost:25", user="report@xxxxxx.com", password="password", dest="report@xxxxxx.com"]
```

**Reply by sebres** [reply]
You did not provide your fail2ban version, so I don't know which config is suitable for you.

You cannot use `%(action_)s` if you don't have `action_` substitution in your `jail.conf` file, like this - https://github.com/fail2ban/fail2ban/blob/884cbbd6e17dd97a87a5c6f7527c0512698ea5d2/config/jail.conf#L212

On some versions you could also use:
```
action    = %(known/action)s
            ...other-actions-with-params...
```
If it also does not work (you have very old version), write the value of `action` specified in `jail.conf` directly.

**Comment by murdocklawless** [comment]
I'm sorry I forgot, my mistake.

fail2ban version v0.11.1 

jail.conf action_ line is exist;

action_ = %(banaction)s[name=%(__name__)s, port="%(port)s", protocol="%(protocol)s", chain="%(chain)s"]

**Comment by murdocklawless** [comment]
I've found an example and tried it and it worked.

[DEFAULT]
destemail = fail2ban@xxxxxx.com
sender = report@xxxxxx.com
mta = mail
action = %(action_mwl)s

edit : mails are coming about jails started or stopped but there is no mail coming about which ip is banned or unbanned?

**Comment by murdocklawless** [comment]
here is my complete jail.local file, maybe it helps you;

these days I usually see named-refused-udp or named-refused-tcp bans in iptables. these two jails has it's own actions, maybe default action and jail action make conflict so because of this ban mails didn't sent?

```
[DEFAULT]
ignoreip = xx.xx.xx.xx
bantime  = 24h
findtime  = 10m
maxretry = 3
destemail = fail2ban@xxxxxx.com
sender = report@xxxxxx.com
mta = mail
action = %(action_mwl)s

[pure-ftpd]
enabled  = true
port     = ftp
filter   = pure-ftpd
logpath  = /var/log/syslog

[dovecot]
enabled = true
filter = dovecot
action = iptables-multiport[name=dovecot-pop3imap, port="pop3,pop3s,imap,imaps", protocol=tcp]
logpath = /var/log/mail.log

[postfix]
enabled  = true
port     = smtp
filter   = postfix
logpath  = /var/log/mail.log

[roundcube-auth]
enabled  = true
port     = 80,443
protocol = tcp
filter   = roundcube-auth
logpath  = /var/log/roundcube/errors.log

[nextcloud]
enabled = true
port = 80,443
protocol = tcp
filter = nextcloud
logpath = /var/log/nextcloud.log

[phpmyadmin-syslog]
enabled = true
port = 80,443
protocol = tcp
filter = phpmyadmin-syslog
logpath = /var/log/auth.log

[ispconfig]
enabled = true
port = 8080
protocol = tcp
filter = ispconfig
logpath = /var/log/ispconfig/auth.log

[named-refused-udp]
enabled   = true
port      = domain,953
protocol  = udp
filter    = named-refused
action    = iptables-multiport[name=named-udp, port="domain,953", protocol=udp]
logpath   = /var/log/syslog

[named-refused-tcp]
enabled   = true
port      = domain,953
protocol  = tcp
filter    = named-refused
action    = iptables-multiport[name=named-tcp, port="domain,953", protocol=tcp]
logpath   = /var/log/syslog

[apache-noscript]
enabled = true
port = 80,443
protocol = tcp
filter = apache-noscript
logpath = /var/log/apache2/error.log

[apache-nohome]
enabled = true
port = 80,443
protocol = tcp
filter = apache-nohome
logpath = /var/log/apache2/error.log

[apache-auth]
enabled = true
port = 80,443
protocol = tcp
filter = apache-auth
logpath = /var/log/apache2/error.log
```

**Reply by sebres** [reply]
> ERROR Failed during configuration: While reading from '/etc/fail2ban/jail.local' [line 63]: option 'action' in section 'name
> jail.conf action_ line is exist

Strange... It should being working with that interpolation... anyway you switched to different config now.

> mails are coming about jails started or stopped but there is no mail coming about which ip is banned or unbanned?

Your mailing action is `mail-whois-lines` now, it must send mails by new bans also (emphasis on new, so restored active bans after restart of fail2ban will be not notified).
If you don't see the banning mails further, take a look in fail2ban.log whether some errors occur after `[jail] Ban 192.0.2.1`.

> maybe default action and jail action make conflict so because of this ban mails didn't sent?

It does not conflict but surely overwrites action parameter in this jails.
Just use `banaction` instead of `action`. And because you have 2 different jails, basically it is enough to set protocol and port in jails, if your banaction is already `iptables-multiport`, otherwise simply set `banaction = iptables-multiport`. 
But if you would combine both jails (-tcp, -udp) to single jail (because iptables-multiport is single protocol action), you can do this:
```ini
[named-refused]
enabled   = true
action_  = %(known/action_)s[name=%(__name__)s-tcp, protocol="tcp"]
           %(known/action_)s[name=%(__name__)s-udp, protocol="udp"]
logpath   = /var/log/syslog
```
This way it would overwrite default `action_` with 2 actions and by interpolation of `action` it result to 3 actions in this jail.
To check resulting configuration you can dump the parameters with `fail2ban-client --dp`.

> here is my complete jail.local file

You don't need to overwrite parameters which all already correctly specified in jail.conf.

```diff
  [dovecot]
  enabled = true
- filter = dovecot
- action = iptables-multiport[name=dovecot-pop3imap, port="pop3,pop3s,imap,imaps", protocol=tcp]
+ banaction = iptables-multiport
+ port = pop3,pop3s,imap,imaps
logpath = /var/log/mail.log
```

**Comment by murdocklawless** [comment]
I just changed action to banaction and mails started to came.

thank you very much for your help.
